cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/DelayedHttpServiceResponder.java

```diff
@@ -155,9 +155,44 @@ public void sendError(int status, String errorMessage) {
    */
   @Override
   public void send(int status, ByteBuffer content, String contentType, Multimap<String, String> headers) {
-    responder.sendContent(HttpResponseStatus.valueOf(status),
-                          ChannelBuffers.wrappedBuffer(content), contentType, headers);
-    emitMetrics(status);
+    send(status, content, contentType, headers, true);
+  }
+
+  private void send(int status, ByteBuffer content, String contentType,
+                    Multimap<String, String> headers, boolean copy) {
+    if (bufferedResponse != null) {
+      LOG.warn("Multiple calls to one of the 'send*' methods has been made. Only the last response will be sent.");
+    }
+
+    ChannelBuffer channelBuffer = null;
+    if (content != null) {
+      channelBuffer = copy ? ChannelBuffers.copiedBuffer(content) : ChannelBuffers.wrappedBuffer(content);
+    }
+
+    bufferedResponse = new BufferedResponse(status, channelBuffer, contentType, headers);
+  }
+
+  /**
+   * Since calling one of the send methods multiple times logs a warning, upon transaction failures this
+   * method is called to allow setting the failure response without an additional warning.
+   */
+  public void setTransactionFailureResponse() {
+    ByteBuffer buffer = Charsets.UTF_8.encode("Transaction failure when committing changes. Aborted transaction.");
+    bufferedResponse = new BufferedResponse(HttpResponseStatus.INTERNAL_SERVER_ERROR.getCode(),
```

Comment:
Yeah, I'm not sure what the right thing to do would be. Merging for now, but we can talk about this further.